### DISCUSSION TITLE
Flask SocketIO CORS Error After Deploy

### DISCUSSION THREAD

**Discussion by selmanbaskaya** [miguelgrinberg/Flask-SocketIO#1762] [Q&A]
ERROR: Access to XMLHttpRequest at 'https://blabla.com/socket.io/...' from origin 'https://blabla.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

I am using Flask SocketIO with VueJs. Everything works fine in my local, but after deploy I started getting CORS error.
I checked the documentation but couldn't find the cause of the error.
** I added cors_allow_origins='*' option for the socket io initialization.

my init.py file is here:
.
.
.
app = Flask(__name__)
CORS(app)
socketio = SocketIO(app, cors_allowed_origins="*")
.
.
.

my main.py file is here: 
.
.
.
if __name__ == "__main__":
    socketio.run(app)

**Comment by miguelgrinberg** [comment]
I think your attempt to obscure your real hostname is complicating this issue, because I assume you are not really using `https://blabla.com`?

The error as you are showing it makes no sense, because it complains about a different origin, while in fact the two URLs shown in the error message have the same origin.

**Reply by selmanbaskaya** [reply]
You are so right it's my fault. Actually, the hostnames are different. It may not be correct to write the real host name of the projects. but as you said, the hostnames are different.

Let the request outgoing hostname be abc.com and the requesting hostname def.com. This is my real scenario. But I don't get CORS error when sending normal request (GET, POST...), I get error when sending emit with socketio.

**I used flask_cors in my projects. Additionally I added this option "cors_allowed_origins='*'" when initializing socket io.

**Reply by miguelgrinberg** [reply]
I'm sorry but I cannot discuss a hypothetical scenario. I don't have any indication to believe the CORS support is broken. If you think it is, then you need to give me actual evidence that I can use to fix the problem.

**Reply by selmanbaskaya** [reply]
I got a few permissions, now I can use the actual evidence.

ERROR: Access to XMLHttpRequest at 'https://twinapi.com/socket.io/...' from origin 'https://twinide.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

While working in my local, I am sending a socketio connection from localhost:8080 to localhost:5001. There is no problem while doing this.

But after deploying, I need to send from twinide.com/twinapi to twinapi.com/api/. when i do this i get the above error.

If I send a request from twinide.com/twinide to twinapi.com/api without socketio, I do not get an error. I get an error when I want to run socketio.

I looked at the deployment documentation of Flask-SocketIO. Everything seems right. I can share if you need any other information.

**Reply by selmanbaskaya** [reply]
Frontend Codes;

const socket = io(process.env.BACKEND_PORT)
socket.on('connect', function () {
socket.emit('username', 'selman');
}.bind(this))


Backend Codes;

app = Flask(name)
CORS(app)
socketio = SocketIO(app, cors_allowed_origins="*")

@socketio.on('username')
def printUsername(username):
print('username is here: ', username)

**Reply by miguelgrinberg** [reply]
@selmanbaskaya I don't really know how to help. If you provide a short and complete example application that shows CORS isn't working I can debug it here, because as I said, all my applications that use CORS settings are working properly.

**Reply by selmanbaskaya** [reply]
Frontend: 

<img width="710" alt="Screen Shot 2022-01-06 at 14 43 27" src="https://user-images.githubusercontent.com/39278203/148378787-f9da1ede-ffc4-46ad-abe6-6c4e68086bb3.png">

Backend app/init.py

<img width="473" alt="Screen Shot 2022-01-06 at 14 46 32" src="https://user-images.githubusercontent.com/39278203/148378822-3309c04e-3295-4a9e-978d-1a519b16e9c9.png">

Backend main.py

<img width="325" alt="Screen Shot 2022-01-06 at 14 47 35" src="https://user-images.githubusercontent.com/39278203/148378845-9175e398-f006-4b7e-a1da-492b7d487bac.png">

If I run these code blocks in my local, I get no errors. When I run it on the server, I get a CORS error

** I'm using supervisor to run gunicorn server. 
** Question: Is there a different configuration I need to do on the server?

**Reply by miguelgrinberg** [reply]
I cannot run and debug screenshots, and even if I could, you are not providing complete front and back end applications that are runnable.

Why don't you take the example in this repository and make the necessary changes to show the bug you are trying to report?

**Reply by selmanbaskaya** [reply]
I created a repo for you to debug more easily.

I wrote in the README how to run Frontend and Backend projects.

Repo: https://github.com/selmanbaskaya/SocketIO-QuasarVue2-Flask

After running the projects, you will see outputs in the browser console and terminal of the backend project.

This code works in my local, does not work on the server. I am getting CORS error.

**Reply by miguelgrinberg** [reply]
There is nothing in your project that can cause the problem you describe. My guess is that some other component in your host is filtering your CORS headers.

**Reply by selmanbaskaya** [reply]
Do I need a different setting for the server? Because as you said, these codes work in my local.

from app import app
from app import socketio

if __name__ == "__main__":
    socketio.run(app)

gunicorn  --workers=3 --threads=2  --timeout 1200000 --bind 0.0.0.0:5001
Isn't it enough to start the server with gunicorn like this?

**Reply by miguelgrinberg** [reply]
How you start the server depends on the async mode that you are using. See the [documentation](https://flask-socketio.readthedocs.io/en/latest/deployment.html#gunicorn-web-server). The `--workers` is probably wrong, and the `--timeout` is unnecessary and might be covering other problems that you have in your application.

**Reply by selmanbaskaya** [reply]
hmm, actually everything works fine with workers and timeout. requests are sent to all other endpoints in the project. I get an error only when I want to connect to socketio.

**Reply by selmanbaskaya** [reply]
![Screen Shot 2022-01-10 at 13 00 27](https://user-images.githubusercontent.com/39278203/148748314-3fe2c5de-6249-427e-8f07-88ee4c4949ee.png)

When I want to connect to socketio, "Access-Control-Allow-Origin:" key does not appear in response header. twinide.com appears in the response header of the other request.

**Reply by miguelgrinberg** [reply]
You have to debug this yourself. You told this library to include the CORS headers in the responses, and you do have this working when you connect directly to your server in your local set up. Something in your production server must be filtering these headers out. What other option is there?

Also, regarding the timeout, it's really none of my business, but having a timeout of 333 hours is definitely not something I would consider normal. The default 30 second timeout in Gunicorn is sufficient in most cases.

**Reply by selmanbaskaya** [reply]
You are right, thank you for answering my questions for 1 week. When I find the solution, I will share it here.

Also, I will consider your suggestion regarding timeout.
